File: BudgetMasterBackend/repository/AccountRepository.java
package repository;

import model.Account;
import util.DateTimeUtil;
import java.sql.*;
import java.util.*;

public class AccountRepository extends BaseRepository implements Repository<Account, Integer> {

    public AccountRepository(String dbPath) {
        super(dbPath);
    }

    @Override
    public Account save(Account account) {
        // Check if there's a deleted record with the same title and restore it
        if (checkAndRestoreDeletedRecord(account, "accounts", "title", account.getTitle())) {
            return account; // Record was restored
        }
        
        // Automatically set position if not set (equals 0)
        setAutoPosition(account, "accounts");
        
        String sql = "INSERT INTO accounts (create_time, update_time, delete_time, created_by, updated_by, deleted_by, position, title, amount, type, currency_id, closed, credit_card_limit, credit_card_category_id, credit_card_commission_category_id) " +
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            
            // Форматируем даты в совместимом с SQLite формате
            String createTimeStr = DateTimeUtil.formatForSqlite(account.getCreateTime());
            String updateTimeStr = DateTimeUtil.formatForSqlite(account.getUpdateTime());
            String deleteTimeStr = DateTimeUtil.formatForSqlite(account.getDeleteTime());
            
            stmt.setString(1, createTimeStr);
            stmt.setString(2, updateTimeStr);
            stmt.setString(3, deleteTimeStr);
            stmt.setString(4, account.getCreatedBy());
            stmt.setString(5, account.getUpdatedBy());
            stmt.setString(6, account.getDeletedBy());
            stmt.setInt(7, account.getPosition());
            stmt.setString(8, account.getTitle());
            stmt.setInt(9, account.getAmount());
            stmt.setInt(10, account.getType());
            stmt.setInt(11, account.getCurrencyId());
            stmt.setInt(12, account.getClosed());
            if (account.getCreditCardLimit() != null) stmt.setInt(13, account.getCreditCardLimit()); else stmt.setNull(13, Types.INTEGER);
            if (account.getCreditCardCategoryId() != null) stmt.setInt(14, account.getCreditCardCategoryId()); else stmt.setNull(14, Types.INTEGER);
            if (account.getCreditCardCommissionCategoryId() != null) stmt.setInt(15, account.getCreditCardCommissionCategoryId()); else stmt.setNull(15, Types.INTEGER);
            stmt.executeUpdate();
            
            // Получаем сгенерированный id
            try (ResultSet rs = stmt.getGeneratedKeys()) {
                if (rs.next()) {
                    account.setId(rs.getInt(1));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        
        // Normalize positions after save
        normalizePositions("accounts");
        
        return account;
    }

    @Override
    public Optional<Account> findById(Integer id) {
        return findById("accounts", id, this::mapRowSafe);
    }

    @Override
    public List<Account> findAll() {
        return findAll("accounts", this::mapRowSafe);
    }

    @Override
    public Account update(Account account) {
        // Adjust positions if needed before updating
        adjustPositionsForUpdate(account, "accounts", account.getId());
        
        String sql = "UPDATE accounts SET create_time=?, update_time=?, delete_time=?, created_by=?, updated_by=?, deleted_by=?, position=?, title=?, amount=?, type=?, currency_id=?, closed=?, credit_card_limit=?, credit_card_category_id=?, credit_card_commission_category_id=? WHERE id=?";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            
            // Форматируем даты в совместимом с SQLite формате
            String createTimeStr = DateTimeUtil.formatForSqlite(account.getCreateTime());
            String updateTimeStr = DateTimeUtil.formatForSqlite(account.getUpdateTime());
            String deleteTimeStr = DateTimeUtil.formatForSqlite(account.getDeleteTime());
            
            stmt.setString(1, createTimeStr);
            stmt.setString(2, updateTimeStr);
            stmt.setString(3, deleteTimeStr);
            stmt.setString(4, account.getCreatedBy());
            stmt.setString(5, account.getUpdatedBy());
            stmt.setString(6, account.getDeletedBy());
            stmt.setInt(7, account.getPosition());
            stmt.setString(8, account.getTitle());
            stmt.setInt(9, account.getAmount());
            stmt.setInt(10, account.getType());
            stmt.setInt(11, account.getCurrencyId());
            stmt.setInt(12, account.getClosed());
            if (account.getCreditCardLimit() != null) stmt.setInt(13, account.getCreditCardLimit()); else stmt.setNull(13, Types.INTEGER);
            if (account.getCreditCardCategoryId() != null) stmt.setInt(14, account.getCreditCardCategoryId()); else stmt.setNull(14, Types.INTEGER);
            if (account.getCreditCardCommissionCategoryId() != null) stmt.setInt(15, account.getCreditCardCommissionCategoryId()); else stmt.setNull(15, Types.INTEGER);
            stmt.setInt(16, account.getId());
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        
        // Normalize positions after update
        normalizePositions("accounts");
        
        return account;
    }

    @Override
    public boolean delete(Integer id) {
        return softDelete("accounts", id, "system");
    }

    /**
     * Soft delete with custom deletedBy parameter
     * @param id entity ID
     * @param deletedBy user who deleted the entity
     * @return true if successful
     */
    public boolean delete(Integer id, String deletedBy) {
        return softDelete("accounts", id, deletedBy);
    }

    /**
     * Restores a soft-deleted account
     * @param id account ID
     * @return true if successful
     */
    public boolean restore(Integer id) {
        return restore("accounts", id);
    }

    /**
     * Gets only deleted accounts
     * @return list of deleted accounts
     */
    public List<Account> findDeleted() {
        return findDeleted("accounts", this::mapRowSafe);
    }

    /**
     * Normalizes positions of all active accounts to be sequential starting from 1
     */
    public void normalizePositions() {
        normalizePositions("accounts");
    }

    private Account mapRow(ResultSet rs) throws SQLException {
        Account account = new Account();
        account.setId(rs.getInt("id"));
        
        // Читаем даты как строки и парсим их
        String createTimeStr = rs.getString("create_time");
        account.setCreateTime(DateTimeUtil.parseFromSqlite(createTimeStr));
        
        String updateTimeStr = rs.getString("update_time");
        account.setUpdateTime(DateTimeUtil.parseFromSqlite(updateTimeStr));
        
        String deleteTimeStr = rs.getString("delete_time");
        account.setDeleteTime(DateTimeUtil.parseFromSqlite(deleteTimeStr));
        
        account.setCreatedBy(rs.getString("created_by"));
        account.setUpdatedBy(rs.getString("updated_by"));
        account.setDeletedBy(rs.getString("deleted_by"));
        account.setPosition(rs.getInt("position"));
        account.setTitle(rs.getString("title"));
        account.setAmount(rs.getInt("amount"));
        account.setType(rs.getInt("type"));
        account.setCurrencyId(rs.getInt("currency_id"));
        account.setClosed(rs.getInt("closed"));
        int val;
        val = rs.getInt("credit_card_limit");
        account.setCreditCardLimit(rs.wasNull() ? null : val);
        val = rs.getInt("credit_card_category_id");
        account.setCreditCardCategoryId(rs.wasNull() ? null : val);
        val = rs.getInt("credit_card_commission_category_id");
        account.setCreditCardCommissionCategoryId(rs.wasNull() ? null : val);
        return account;
    }

    private Account mapRowSafe(ResultSet rs) {
        try {
            return mapRow(rs);
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }
} 

File: BudgetMasterBackend/examples/repository/account_repository.py
import os
import sys

from BudgetMasterBackend.examples.common import (
    cleanup_example,
    create_test_entity,
    get_java_class,
    setup_example,
    test_data_manager,
)

sys.path.append(os.path.dirname(os.path.dirname(__file__)))


def main():
    print("=== Тест AccountRepository через JPype ===")

    # Настройка окружения
    if not setup_example():
        return

    try:
        # Импортируем классы
        Account = get_java_class("model.Account")
        AccountRepository = get_java_class("repository.AccountRepository")
        LocalDateTime = get_java_class("java.time.LocalDateTime")

        print("✅ Классы импортированы")

        # Создаем репозиторий
        repo = AccountRepository(test_data_manager.db_manager.db_path)
        print("✅ Репозиторий создан")

        # Создаем тестовый счет
        account = create_test_entity(
            Account,
            title="Наличные (тест)",
            amount=100000,  # 1000.00 в копейках
            type=1,  # Текущий счет
            currencyId=1,  # RUB
            closed=0,  # Не закрыт
            createdBy="tester",
            updatedBy="tester",
        )

        print("✅ Тестовый счет создан")
        print(f"Счет: {account.toString()}")

        # Тестируем сохранение
        print("\n--- Тест сохранения ---")
        saved_account = repo.save(account)
        print(f"Счет сохранен: {saved_account.toString()}")
        print(f"ID счета: {saved_account.getId()}")
        test_data_manager.add_test_id("accounts", saved_account.getId())

        # Тестируем поиск по ID
        print("\n--- Тест поиска по ID ---")
        try:
            found_account = repo.findById(saved_account.getId())
            if found_account.isPresent():
                acc = found_account.get()
                print(f"Найден счет: {acc.toString()}")
                print(f"ID: {acc.getId()}")
                print(f"Position: {acc.getPosition()}")
                print(f"Title: {acc.getTitle()}")
                print(f"Amount: {acc.getAmount()} копеек")
                print(f"Type: {acc.getType()}")
                print(f"Currency ID: {acc.getCurrencyId()}")
                print(f"Closed: {acc.getClosed()}")
                print(f"Created by: {acc.getCreatedBy()}")
            else:
                print("❌ Счет не найден")
        except Exception as e:
            print(f"❌ Ошибка при поиске: {e}")
            import traceback
            traceback.print_exc()

        # Прямой SQL-запрос для проверки
        print("\n--- Прямой SQL-запрос ---")
        try:
            import sqlite3

            conn = sqlite3.connect(test_data_manager.db_manager.db_path)
            cursor = conn.cursor()

            cursor.execute("SELECT * FROM accounts WHERE id = ?", (saved_account.getId(),))
            row = cursor.fetchone()

            if row:
                print("✅ Найден счет в БД:")
                cursor.execute("PRAGMA table_info(accounts)")
                columns = [col[1] for col in cursor.fetchall()]

                for i, (col_name, value) in enumerate(zip(columns, row)):
                    print(f"  {col_name}: {value} (тип: {type(value).__name__})")
            else:
                print("❌ Счет не найден в БД")

            conn.close()

        except Exception as e:
            print(f"❌ Ошибка при прямом SQL-запросе: {e}")

        # Тестируем обновление
        print("\n--- Тест обновления ---")
        try:
            saved_account.setTitle("Основной счет (обновленный)")
            saved_account.setPosition(2)
            saved_account.setAmount(150000)  # 1500.00 в копейках
            saved_account.setUpdatedBy("jpype_update_test")
            saved_account.setUpdateTime(LocalDateTime.now())

            updated_account = repo.update(saved_account)
            print(f"Счет обновлен: {updated_account.toString()}")

            found_updated = repo.findById(saved_account.getId())
            if found_updated.isPresent():
                acc = found_updated.get()
                print(f"Проверка обновления: {acc.toString()}")
                print(f"Новый title: {acc.getTitle()}")
                print(f"Новый position: {acc.getPosition()}")
                print(f"Новый amount: {acc.getAmount()} копеек")
                print(f"Updated by: {acc.getUpdatedBy()}")
            else:
                print("❌ Обновленный счет не найден")
        except Exception as e:
            print(f"❌ Ошибка при обновлении: {e}")
            import traceback
            traceback.print_exc()

        # Тестируем soft delete
        print("\n--- Тест soft delete ---")
        try:
            account_id = saved_account.getId()
            all_before_delete = repo.findAll()
            print(f"Счетов до удаления: {len(all_before_delete)}")

            deleted = repo.delete(account_id, "test_user")
            print(f"Счет помечен как удаленный: {deleted}")

            all_after_delete = repo.findAll()
            print(f"Счетов после удаления (обычный режим): {len(all_after_delete)}")

            repo.setIncludeDeleted(True)
            all_with_deleted = repo.findAll()
            print(f"Счетов после удаления (включая удаленные): {len(all_with_deleted)}")

            found_deleted = repo.findById(account_id)
            if found_deleted.isPresent():
                deleted_account = found_deleted.get()
                print(f"Найден удаленный счет: {deleted_account.getTitle()}")
                print(f"Deleted by: {deleted_account.getDeletedBy()}")
                print(f"Delete time: {deleted_account.getDeleteTime()}")
            else:
                print("❌ Удаленный счет не найден")

            # Восстановление через restore
            print("\n--- Тест восстановления через restore ---")
            restored = repo.restore(account_id)
            print(f"Счет восстановлен через restore: {restored}")

            all_after_restore = repo.findAll()
            print(f"Счетов после восстановления: {len(all_after_restore)}")

            found_restored = repo.findById(account_id)
            if found_restored.isPresent():
                restored_account = found_restored.get()
                print(f"Найден восстановленный счет: {restored_account.getTitle()}")
                print(f"Delete time после восстановления: {restored_account.getDeleteTime()}")
                print(f"Deleted by после восстановления: {restored_account.getDeletedBy()}")
            else:
                print("❌ Восстановленный счет не найден")

        except Exception as e:
            print(f"❌ Ошибка при soft delete: {e}")
            import traceback
            traceback.print_exc()

        print("\n✅ Все тесты выполнены успешно!")

    except Exception as e:
        print(f"❌ Ошибка: {e}")
        import traceback
        traceback.print_exc()

    finally:
        cleanup_example()


if __name__ == "__main__":
    main()
